workflow/controller/controller.go

```diff
@@ -523,32 +531,49 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 	logCtx.Info("cleaning up pod")
 	err := func() error {
 		pods := wfc.kubeclientset.CoreV1().Pods(namespace)
+		pod, err := pods.Get(ctx, podName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
 		switch action {
 		case terminateContainers:
-			pod, err := wfc.getPod(namespace, podName)
-			if err == nil && pod != nil && pod.Status.Phase == apiv1.PodPending {
-				wfc.queuePodForCleanup(namespace, podName, deletePod)
-			} else if terminationGracePeriod, err := wfc.signalContainers(namespace, podName, syscall.SIGTERM); err != nil {
+			patch := createFinalizerPodStatusRemovalPatchIfExists(pod)
+			if err := applyPatchIfExists(ctx, pods, podName, patch); err != nil {
 				return err
-			} else if terminationGracePeriod > 0 {
-				wfc.queuePodForCleanupAfter(namespace, podName, killContainers, terminationGracePeriod)
+			}
+			if pod.Status.Phase == apiv1.PodPending {
+				wfc.queuePodForCleanup(namespace, podName, deletePod)
+			} else {
+				terminationGracePeriod := wfc.signalContainers(pod, syscall.SIGTERM)
+				if terminationGracePeriod > 0 {
+					wfc.queuePodForCleanupAfter(namespace, podName, killContainers, terminationGracePeriod)
+				}
 			}
 		case killContainers:
-			if _, err := wfc.signalContainers(namespace, podName, syscall.SIGKILL); err != nil {
+			patch := createFinalizerPodStatusRemovalPatchIfExists(pod)
+			if err := applyPatchIfExists(ctx, pods, podName, patch); err != nil {
 				return err
 			}
+			wfc.signalContainers(pod, syscall.SIGKILL)
```

Comment:
Can I understand the rationale for doing this from `terminateContainers` and `killContainers`? (I'm just guessing `deletePod` would be called later in those cases?)

Reply:
From my research, `terminateContainers` appears to be used [here](https://github.com/argoproj/argo-workflows/blob/v3.5.2/workflow/controller/exec_control.go#L48), for example.
As far as I know, `killContainers` is only used in the `case terminateContainers` section.

`deletePod` calls `pods.Delete()`.

Reply:
So, I guess my question was whether it's redundant to both remove the finalizer in "terminateContainers" as well as in "deletePod"? In the line you pointed to, won't the "terminateContainers" logic ultimately cause the "deletePod" logic? (I'm also willing to entertain the idea that we do it in both places just in case, but want to understand.)

Reply:
As you pointed out, the finalizer removal logic in `terminateContainers` is indeed redundant upon further review. Thank you for highlighting this. I have removed the redundant finalizer removal logic in the commit below.
https://github.com/argoproj/argo-workflows/pull/12413/commits/b73ed375a03210c1fec4045894b28ccb2b6e1841

Reply:
Can you check if `killContainers` is the same?

Reply:
I believe it's the same and it was removed as well.
Additionally, I also added the executor e2e test with the finalizer option set to true, just to be safe.
https://github.com/argoproj/argo-workflows/pull/12413/commits/cf831c06b7756d901f3e76644fa019c3ab5ab2f5